benchmarks/filters/voxel_grid.cpp

```diff
@@ -1,71 +1,92 @@
 
+#include <boost/filesystem.hpp> // std implementation from C++17
+
 #include <benchmark/benchmark.h>
 
+#include <string>
+#include <vector>
+
+using namespace pcl;
+namespace fs = boost::filesystem;
+
+template <typename PointT, template <typename> class GridFilter>
 static void
-BM_VoxelGrid(benchmark::State& state, const std::string& file)
+BM_grid_filter(benchmark::State& state, const std::string& file)
 {
   // Perform setup here
-  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
-  pcl::PCDReader reader;
+  typename PointCloud<PointT>::Ptr cloud(new PointCloud<PointT>);
+  PCDReader reader;
   reader.read(file, *cloud);
 
-  pcl::VoxelGrid<pcl::PointXYZ> vg;
-  vg.setLeafSize(0.01, 0.01, 0.01);
-  vg.setInputCloud(cloud);
+  GridFilter<PointT> f;
+  f.setLeafSize(0.01, 0.01, 0.01);
+  f.setInputCloud(cloud);
 
-  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_voxelized(
-      new pcl::PointCloud<pcl::PointXYZ>);
+  PointCloud<PointT> cloud_voxelized;
   for (auto _ : state) {
     // This code gets timed
-    vg.filter(*cloud_voxelized);
+    f.filter(cloud_voxelized);
   }
 }
 
-static void
-BM_ApproxVoxelGrid(benchmark::State& state, const std::string& file)
-{
-  // Perform setup here
-  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
-  pcl::PCDReader reader;
-  reader.read(file, *cloud);
+template <typename PointT, template <typename> class... Ts>
+struct BM_register;
 
-  pcl::ApproximateVoxelGrid<pcl::PointXYZ> avg;
-  avg.setLeafSize(0.01, 0.01, 0.01);
-  avg.setInputCloud(cloud);
+template <typename PointT>
+struct BM_register<PointT> {
+  BM_register(const std::vector<std::string>::const_iterator, const std::string&) {}
+};
 
-  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_voxelized(
-      new pcl::PointCloud<pcl::PointXYZ>);
-  for (auto _ : state) {
-    // This code gets timed
-    avg.filter(*cloud_voxelized);
+template <typename PointT, template <typename> class T, template <typename> class... Ts>
+struct BM_register<PointT, T, Ts...> {
+  BM_register(const std::vector<std::string>::const_iterator class_name,
+              const std::string& dataset)
+  {
+    const std::string bm_name = fs::path(dataset).stem().string() + "_" + *class_name;
+    benchmark::RegisterBenchmark(bm_name.c_str(), &BM_grid_filter<PointT, T>, dataset)
+        ->Unit(benchmark::kMillisecond);
+
+    BM_register<PointT, Ts...>(class_name + 1, dataset);
   }
-}
+};
+
+template <typename PointT, template <typename> class... Ts>
+struct BM_registration {
+  BM_registration(const std::vector<std::string>& classes_name,
+                  const std::vector<std::string>& datasets)
+  {
+    if (sizeof...(Ts) != classes_name.size()) {
+      PCL_ERROR(
+          "[Benchmark] Number of template classes and classes name are different.\n");
+      return;
+    }
+
+    for (const auto& dataset : datasets)
+      BM_register<PointT, Ts...>(classes_name.begin(), dataset);
```

Comment:
Did you try Boost MPL or lambdas to iterate?
```cpp
template <class F, class... Args>
void do_for(F f, Args... args) {
    int x[] = {(f(args), 0)...};
}

template <class... Args>
void doSomething(Args... args) {
    do_for([&](auto arg) {
        // You can do something with arg here.
    }, args...);
}
```

This could keep the logic together instead of being split. See https://stackoverflow.com/a/37227316 for more methods

Reply:
Done, I used initialization list to avoid recursion following the solution
(can't use lambda because we're passing types but not objects)